lib/internal/fs/recursive_watch.js

```diff
@@ -0,0 +1,288 @@
+'use strict';
+
+const {
+  ArrayPrototypePush,
+  SafePromiseAll,
+  Promise,
+  PromisePrototypeThen,
+  SafeMap,
+  StringPrototypeStartsWith,
+  StringPrototypeSlice,
+  StringPrototypeLastIndexOf,
+  SymbolAsyncIterator,
+} = primordials;
+
+const { EventEmitter } = require('events');
+const assert = require('internal/assert');
+const {
+  AbortError,
+  codes: {
+    ERR_INVALID_ARG_VALUE,
+  },
+} = require('internal/errors');
+const { getValidatedPath } = require('internal/fs/utils');
+const { kFSWatchStart, StatWatcher } = require('internal/fs/watchers');
+const { kEmptyObject } = require('internal/util');
+const { validateBoolean, validateAbortSignal } = require('internal/validators');
+const path = require('path');
+
+let internalSync;
+let internalPromises;
+
+function lazyLoadFsPromises() {
+  internalPromises ??= require('fs/promises');
+  return internalPromises;
+}
+
+function lazyLoadFsSync() {
+  internalSync ??= require('fs');
+  return internalSync;
+}
+
+async function traverse(dir, files = new SafeMap()) {
+  const { opendir } = lazyLoadFsPromises();
+
+  const filenames = await opendir(dir);
+  const subdirectories = [];
+
+  for await (const file of filenames) {
+    const f = path.join(dir, file.name);
```

Comment:
Shall we use path.resolve for '../../filename.ext' cases?

Reply:
I don't think this can happen, `path.resolve` would already have been called at this point https://github.com/nodejs/node/blob/edee6c5227f0610db30f3684b3a0d0a4ce666fbc/lib/internal/fs/recursive_watch.js#L212

Reply:
If the `file.name` is `../../filename.ext` path.join would resolve to one of the `dir` file's parent folders.

Reply:
I don't think there's any platform (that Node.js supports) that supports `/` in file names.

Reply:
@nodejs/fs

Reply:
I can see that output from Node v19.0.0:

```
c² ~/ [main] node
Welcome to Node.js v19.0.0.
Type ".help" for more information.
> const path = require('path')
undefined
> path.resolve('', '../../../foo/../cagatay/bar.ext')
'/Users/cagatay/bar.ext'
> path.resolve('', '../../bar.ext')
'/Users/cagatay/bar.ext'
>
```

Reply:
Oh, the limitation is system level, @anonrig feel free to resolve, thanks!